### PR TITLE
Presets management

### PR DESCRIPTION
**Pull request by hanazarraa**

### CODE REVIEW

File: 20220511140915_create_preset_category_table.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+     * Remember to call "create()" or "update()" and NOT "save()" when working
+     * with the Table class.
+     */
+    public function change(): void
+    {
+        $table = $this->table('preset_categories');
+        $table
+            ->addColumn('name', 'string', ['limit' => 64, 'null' => false])
+            ->addColumn('enabled','boolean', ['default' => false, 'null' => false])
+            ->addColumn('icon',"string",["null"=>true])
```
Comment (GhaziTriki) [inline 20220511140915_create_preset_category_table.php:25]: I don't think we need the icon name here. It should be guessable from the category name. Convention over configuration.

File: 20220511140915_create_preset_category_table.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+     *
+     * Write your reversible migrations using this method.
+     *
+     * More information on writing migrations is available here:
+     * https://book.cakephp.org/phinx/0/en/migrations.html#the-change-method
+     *
+     * Remember to call "create()" or "update()" and NOT "save()" when working
+     * with the Table class.
+     */
+    public function change(): void
```
Comment (GhaziTriki) [inline 20220511140915_create_preset_category_table.php:19]: I think we are using `up` by convention.

File: hivelvet-backend/app/src/Actions/Presets/Add.php [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            if (str_starts_with($className, 'ComposerAutoloaderInit')) {
+                $autoloaderClassName = $className;
+
+                break;
+            }
+        }
+        $classLoader = $autoloaderClassName::getLoader();
+        $classMap    = $classLoader->getClassMap();
+
+        $categories = preg_filter('/^Enum\\\Presets\\\[A-Z a-z]*/', '$0', array_keys($classMap));
```
Comment (GhaziTriki) [inline hivelvet-backend/app/src/Actions/Presets/Add.php:82]: Looks like a Titan work :rocket: !! Well done. However can you help simple human understand it by adding comment lines?

File: hivelvet-backend/app/src/Enum/Presets/Audio.php [hunk 4]
```diff
@@ -32,5 +32,10 @@ class Audio extends Enum
     final public const LISTEN_ONLY                        = 'listen_only_enabled';
     final public const SKIP_ECHO_TEST                     = 'skip_echo_test';
 
-    public static string $icon = 'AudioOutlined';
+    public static string $icon                                    = 'AudioOutlined';
```
Comment (GhaziTriki) [inline hivelvet-backend/app/src/Enum/Presets/Audio.php:35]: I don't think we need the icon name here. We can find it by matching the class name.

File: hivelvet-backend/app/src/Models/SubCategoryPreset.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    protected $table = 'preset_subcategories';
+
+    public function findByPresetAndSubCategory($preset_id, $subcategory_id)
+    {
+        $this->load(['sub_category_id = ? && preset_id = ?', $subcategory_id, $preset_id]);
+
+        return $this;
+    }
+
+    public function PresetAndSubCategoryExists($preset_id, $subcategory_id)
```
Comment (GhaziTriki) [inline hivelvet-backend/app/src/Models/SubCategoryPreset.php:48]: Method name starts by capital letter.

**Review by GhaziTriki — CHANGES_REQUESTED** [review]
The name of preset category, subcategory is confusing.
By just checking the UI I can see that we have:
- Preset that has a name.
- Preset has multiple "Configuration Group".
- Each configuration group has a "Configuration Entry".
Did I miss anything else?

File: hivelvet-backend/app/src/Actions/Presets/Delete.php [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            try {
+                $preset->erase();
+            } catch (\Exception $e) {
+                $message = 'preset  could not be deleted';
+                $this->logger->error('preset could not be deleted', ['preset' => $preset->toArray(), 'error' => $e->getMessage()]);
+                $this->renderJson(['message' => $message], ResponseCode::HTTP_INTERNAL_SERVER_ERROR);
+
+                return;
+            }
+            $this->logger->info('preset successfully deleted');
```
Comment (GhaziTriki) [inline hivelvet-backend/app/src/Actions/Presets/Delete.php:52]: We need to print the preset id.

**Review by GhaziTriki — APPROVED** [review]

**Comment by SaoussenBlk** [comment]
#87 #51  #73

**Pull request merged**